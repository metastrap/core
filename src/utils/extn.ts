// eslint-disable-next-line import/prefer-default-export
export function getExtn(filename: string) {
  return /\.(\w+)$/.exec(filename)?.[1];
}
