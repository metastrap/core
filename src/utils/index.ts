export * as astUtils from './ast';
export * as logger from './logger';
export * as read from './read';
