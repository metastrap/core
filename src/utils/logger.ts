/* eslint-disable no-console */
export default function logger(...args: unknown[]) {
  console.log(...args);
}
