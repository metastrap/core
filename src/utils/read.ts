import JSZip from 'jszip';

export default async function readZipToMap(
  zip: JSZip,
): Promise<Map<string, string>> {
  const files = Object.keys(zip.files);

  const maybeContentList = files.map(async (file) => {
    if (!zip.files[file].dir) {
      const contents = await zip.files[file].async('text');
      return [
        file,
        contents || '',
      ] as [string, string];
    }
    /* directories will return null */
    return null;
  });

  const contentList = (
    await Promise.all(maybeContentList)
  ).filter((file) => !!file);

  return new Map(contentList);
}
