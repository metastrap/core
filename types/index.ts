import type { ParseResult } from '@babel/parser';
import type { File } from '@babel/types';

export type TAstValue = string | ParseResult<File> | { [key: string]: unknown };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type TObject = { [key: string]: any };
