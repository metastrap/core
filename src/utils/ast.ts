/* browser */
import { ParseResult, ParserPlugin, parse } from '@babel/parser';
import generate from '@babel/generator';

import type { File } from '@babel/types';

import { TAstMap, TAstValue } from 'types';
import logger from './logger';

enum EFileFeature {
  JAVASCRIPT = 1,
  TYPESCRIPT = 2,
  JSX = 4,
  JSON = 8,
}

export function convertTextToAst(
  fileContentMap: Map<string, string>,
): TAstMap {
  const resultMap = new Map<string, TAstValue>();

  fileContentMap.forEach((content, file) => {
    let featureFlag = 0;
    const extname = file.split('.').pop() || '';
    const babelParsePlugins: ParserPlugin[] = ['classProperties'];

    let result: TAstValue = content;
    switch (extname) {
      case 'js':
        featureFlag |= EFileFeature.JAVASCRIPT;
        break;
      case 'ts':
        featureFlag |= EFileFeature.TYPESCRIPT;
        break;
      case 'jsx':
        featureFlag |= EFileFeature.JSX;
        break;
      case 'tsx':
        featureFlag |= EFileFeature.TYPESCRIPT | EFileFeature.JSX;
        break;
      case 'json':
        try {
          result = JSON.parse(content);
        } catch (e) {
          logger((e as Error).message);
        }
    }

    if (featureFlag & EFileFeature.TYPESCRIPT) babelParsePlugins.push('typescript');
    if (featureFlag & EFileFeature.JSX) babelParsePlugins.push('jsx');

    if (featureFlag & (EFileFeature.JAVASCRIPT | EFileFeature.TYPESCRIPT | EFileFeature.JSX)) {
      try {
        result = parse(content, {
          sourceType: 'module',
          plugins: babelParsePlugins,
        });
      } catch (e) {
        logger((e as Error).message);
        result = content;
      }
    }
    resultMap.set(file, result);
  });
  return resultMap;
}

export function print(ast: Map<string, TAstValue>): Map<string, string> {
  const resultMap = new Map<string, string>();
  ast.forEach((astValue, file) => {
    if (typeof astValue === 'string') {
      resultMap.set(file, astValue);
      return;
    }
    if (astValue.program) {
      resultMap.set(file, generate(astValue as ParseResult<File>).code);
      return;
    }
    if (typeof astValue === 'object') {
      try {
        resultMap.set(file, JSON.stringify(astValue, null, 2));
      } catch (e) {
        resultMap.set(file, astValue.toString?.() || '');
      }
    }
  });
  return resultMap;
}
