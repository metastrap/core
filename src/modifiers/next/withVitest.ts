import { addToBaseAst } from '@/utils/ast';
import {
  TAstMap, TObject,
  TROAstMap,
} from 'types';

export const NAMESPACE = 'vitest';

export const vitestFiles = [
  'vitest/__tests__/components/counter.test.jsx',
  'vitest/__tests__/utils/add.test.js',
  'vitest/__tests__/page.test.jsx',
  'vitest.config.js',
  'app/utils/add.js',
  'app/counter.jsx',
];

export default function withVitest(
  baseAstMap: TAstMap,
  fullAstMap: TROAstMap,
) {
  /* package.json */
  const pkg = baseAstMap.get('package.json') as TObject;
  pkg.devDependencies = {
    ...pkg.devDependencies || {},
    '@testing-library/react': '14.1.2',
    '@vitejs/plugin-react': '^4.2.1',
    vitest: '1.0.4',
    jsdom: '^23.0.1',
  };

  pkg.scripts = {
    ...pkg.scripts || {},
    test: 'vitest',
  };

  addToBaseAst(baseAstMap, vitestFiles, fullAstMap);

  /* Rename files since `vitest/` was added for namespacing */
  for (let i = 0; i < 3; i++) {
    const oldKey = vitestFiles[i];
    const newKey = oldKey.replace(`${NAMESPACE}/`, '');
    baseAstMap.set(newKey, baseAstMap.get(oldKey) as TObject);
    baseAstMap.delete(oldKey);
  }
}
