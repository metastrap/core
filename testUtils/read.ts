// eslint-disable-next-line import/no-extraneous-dependencies
import fg from 'fast-glob';
import path from 'node:path';
import fs from 'node:fs/promises';
import { EFrameworks } from '@/constants/enum';
import { TAstValue } from 'types/index';
import { convertTextToAst } from '@/utils/ast';
import { setBaseProject } from '@/project';
import { baseFiles } from '@/modifiers/next';

export async function readFilesToMap(
  dir: string,
): Promise<Map<string, string>> {
  const files = await fg([`${dir}/**/*`], { dot: true });
  return new Map(
    await Promise.all(
      files.map(async (file) => {
        const contents = await fs.readFile(file, 'utf8');
        return [
          file.replace(`${dir}/`, ''),
          contents || '',
        ] as [string, string];
      }),
    ),
  );
}

export async function readFixture(framework: EFrameworks): Promise<Map<string, string>> {
  return await readFilesToMap(
    path.join(process.cwd(), 'fixtures', framework.toString()),
  );
}

export async function astFromFixture(): Promise<[
  Map<string, TAstValue>,
  Map<string, TAstValue>
]> {
  const fullProject = convertTextToAst(
    await readFilesToMap(
      path.join(process.cwd(), 'fixtures', 'next'),
    ),
  );
  return [
    fullProject,
    setBaseProject(fullProject, baseFiles),
  ];
}
