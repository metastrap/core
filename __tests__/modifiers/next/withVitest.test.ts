import { TAstValue } from 'types/index';
import { print } from '@/utils/ast';
import { astFromFixture } from 'testUtils/read';
import withVitest, { NAMESPACE, vitestFiles } from '@/modifiers/next/withVitest';

let baseAsTree: Map<string, TAstValue>;
let fullProject: Map<string, TAstValue>;

describe('modifiers/next/withVitest', () => {
  beforeEach(async () => {
    [fullProject, baseAsTree] = await astFromFixture();
  });

  test('contains required devDependencies', () => {
    withVitest(baseAsTree, fullProject);

    expect(
      print(baseAsTree).get('package.json'),
    ).toMatchSnapshot();
  });

  test('it contains new list of files for vitest', () => {
    withVitest(baseAsTree, fullProject);
    expect(
      vitestFiles.map((file) => typeof baseAsTree.get(file.replace(`${NAMESPACE}/`, ''))),
    ).toStrictEqual(
      new Array(vitestFiles.length)
        .fill('object'),
    );
  });
});
