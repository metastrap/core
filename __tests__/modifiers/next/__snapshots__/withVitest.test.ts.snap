// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifiers/next/withVitest contains required devDependencies 1`] = `
"{
  "name": "hello",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "test": "vitest"
  },
  "dependencies": {
    "react": "^18",
    "react-dom": "^18",
    "next": "14.0.4"
  },
  "devDependencies": {
    "@testing-library/react": "14.1.2",
    "@vitejs/plugin-react": "^4.2.1",
    "vitest": "1.0.4",
    "jsdom": "^23.0.1"
  }
}"
`;
